mlir/lib/Analysis/Presburger/Matrix.cpp

```diff
@@ -276,23 +270,25 @@ void Matrix<T>::moveColumns(unsigned srcPos, unsigned num, unsigned dstPos) {
   assert(dstPos + num <= getNumColumns() &&
          "move destination range exceeds matrix columns");
 
-  unsigned insertCount = offset > 0 ? offset : -offset;
-  unsigned finalAdjStart = offset > 0 ? srcPos : srcPos + num;
-  unsigned curAdjStart = offset > 0 ? srcPos + num : dstPos;
-  // TODO: This can be done using std::rotate.
-  // Insert new zero columns in the positions where the adjacent columns are to
-  // be moved.
-  insertColumns(finalAdjStart, insertCount);
-  // Update curAdjStart if insertion of new columns invalidates it.
-  if (finalAdjStart < curAdjStart)
-    curAdjStart += insertCount;
-
-  // Swap the adjacent columns with inserted zero columns.
-  for (unsigned i = 0; i < insertCount; ++i)
-    swapColumns(finalAdjStart + i, curAdjStart + i);
-
-  // Delete the now redundant zero columns.
-  removeColumns(curAdjStart, insertCount);
+  unsigned numRows = getNumRows();
+  unsigned numCols = getNumReservedColumns();
+
+  if (offset > 0) {
+    // shift the matrix left, see
+    // https://en.cppreference.com/w/cpp/algorithm/rotate.html.
+    for (unsigned i = 0; i < numRows; ++i) {
+      auto begin = data.begin() + i * numCols + srcPos;
+      auto end = data.begin() + i * numCols + dstPos + num;
+      std::rotate(begin, begin + num, end);
+    }
+  } else {
+    // shift the matrix right.
+    for (unsigned i = 0; i < numRows; ++i) {
+      auto begin = data.begin() + i * numCols + srcPos + num;
+      auto end = data.begin() + i * numCols + dstPos;
```

Comment:
the names start and end are confusing here. I would just directly use &at(...) as mentioned above, and write a comment for each case explaining what's happening. Thanks!